File: Leetcode 3501-3600/3548/3548.py
# Leetcode 3548: Equal Sum Grid Partition II
# https://leetcode.com/problems/equal-sum-grid-partition-ii/
# Solved on 24th of November, 2025
from collections import Counter


class Solution:
    def canPartitionGrid(self, grid: list[list[int]]) -> bool:
        """
        Determines if a grid can be partitioned into two subgrids with equal sums by making at most two cuts.

        Args:
            grid: A list of lists of integers representing the grid.
        Returns:
            True if the grid can be partitioned into two subgrids with equal sums, False otherwise.
        """
        def checkCuts(currentGrid):
            m, n = len(currentGrid), len(currentGrid[0])
            bottomCounts = Counter()
            totalSum = 0

            for r in range(m):
                for c in range(n):
                    val = currentGrid[r][c]
                    bottomCounts[val] += 1
                    totalSum += val

            topCounts = Counter()
            topSum = 0

            for i in range(m - 1):
                rowSum = 0
                for c in range(n):
                    val = currentGrid[i][c]
                    topCounts[val] += 1
                    bottomCounts[val] -= 1
                    if bottomCounts[val] == 0:
                        del bottomCounts[val]
                    rowSum += val

                topSum += rowSum
                bottomSum = totalSum - topSum
                diff = topSum - bottomSum

                if diff == 0:
                    return True

                if diff > 0:
                    target = diff
                    if n == 1:
                        if currentGrid[0][0] == target or currentGrid[i][0] == target:
                            return True
                    elif i == 0:
                        if currentGrid[0][0] == target or currentGrid[0][n - 1] == target:
                            return True
                    else:
                        if topCounts[target] > 0:
                            return True
                else:
                    target = -diff
                    if n == 1:
                        if currentGrid[i + 1][0] == target or currentGrid[m - 1][0] == target:
                            return True
                    elif i == m - 2:
                        if currentGrid[m - 1][0] == target or currentGrid[m - 1][n - 1] == target:
                            return True
                    else:
                        if bottomCounts[target] > 0:
                            return True
            return False

        if checkCuts(grid):
            return True

        transposedGrid = [list(row) for row in zip(*grid)]
        if checkCuts(transposedGrid):
            return True

        return False

File: Leetcode 3501-3600/3548/3548.java
// Leetcode 3548: Equal Sum Grid Partition II
// https://leetcode.com/problems/equal-sum-grid-partition-ii/
// Solved on 24th of November, 2025
class Solution {
    /**
     * Determines if a grid can be partitioned into two sub-grids with equal sums by removing a single row or column,
     * and then adjusting one of the sub-grids by adding or removing a single element to make their sums equal.
     * @param grid The input 2D integer array representing the grid.
     * @return True if such a partition is possible, false otherwise.
     */
    public boolean canPartitionGrid(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        long totalSum = 0;
        int[] totalCounts = new int[100001];

        for (int[] row : grid) {
            for (int val : row) {
                totalSum += val;
                totalCounts[val]++;
            }
        }

        int[] part1Counts = new int[100001];
        int[] part2Counts = totalCounts.clone();
        long sum1 = 0;

        for (int i = 0; i < m - 1; i++) {
            for (int j = 0; j < n; j++) {
                int val = grid[i][j];
                sum1 += val;
                part1Counts[val]++;
                part2Counts[val]--;
            }

            long sum2 = totalSum - sum1;
            long diff = Math.abs(sum1 - sum2);

            if (diff == 0) {
                return true;
            }
            if (diff > 100000) {
                continue;
            }

            int target = (int) diff;

            if (sum1 > sum2) {
                if (part1Counts[target] > 0) {
                    if ((i + 1 > 1 && n > 1) || checkEnds(grid, 0, i, 0, n - 1, target)) return true;
                }
            } else {
                if (part2Counts[target] > 0) {
                    if ((m - 1 - i > 1 && n > 1) || checkEnds(grid, i + 1, m - 1, 0, n - 1, target)) return true;
                }
            }
        }

        part1Counts = new int[100001];
        part2Counts = totalCounts.clone();
        sum1 = 0;

        for (int j = 0; j < n - 1; j++) {
            for (int i = 0; i < m; i++) {
                int val = grid[i][j];
                sum1 += val;
                part1Counts[val]++;
                part2Counts[val]--;
            }

            long sum2 = totalSum - sum1;
            long diff = Math.abs(sum1 - sum2);

            if (diff == 0) {
                return true;
            }
            if (diff > 100000) {
                continue;
            }

            int target = (int) diff;

            if (sum1 > sum2) {
                if (part1Counts[target] > 0) {
                    if ((m > 1 && j + 1 > 1) || checkEnds(grid, 0, m - 1, 0, j, target)) {
                        return true;
                    }
                }
            } else {
                if (part2Counts[target] > 0) {
                    if ((m > 1 && n - 1 - j > 1) || checkEnds(grid, 0, m - 1, j + 1, n - 1, target)) {
                        return true;
                    }
                }
            }
        }

        return false;
    }

    private boolean checkEnds(int[][] grid, int rowStart, int rowEnd, int colStart, int colEnd, int target) {
        if (rowStart == rowEnd) {
            return grid[rowStart][colStart] == target || grid[rowStart][colEnd] == target;
        }
        if (colStart == colEnd) {
            return grid[rowStart][colStart] == target || grid[rowEnd][colStart] == target;
        }
        return false;
    }
}